### PR TITLE
updating clxfb and render

### LINKED ISSUE #199: undefined function %make-iamge-sheet-get-function called in render/design.lisp
It looks like there used to be a function called %make-image-sheet-get-function and that it was removed and uses of it replaced with direct (lambda (x y) ...) calls in the various make-rgba-design methods. The method for image-sheet-mixin still calls this now non-existent function. @gas2serra can you take a look at this?

### PR DESCRIPTION
**Pull request by gas2serra**
An improved version of the render extension and of the clxfb backend. 
It fixes #199.
Regards,
Alessandro

### CODE REVIEW

**Comment by slyrus** [comment]
Wow. Big commit. What's the best way to test this?

**Comment by dkochmanski** [comment]
@slyrus regarding checking fb backend, see wiki entry: https://github.com/robert-strandh/McCLIM/wiki/CLX-backend(s)

**Comment by gas2serra** [comment]
(ql:quickload 'clim-examples)
(ql:quickload 'mcclim-clx-fb)
(let ((clim:*default-server-path*  '(:clx-fb)))
    (CLIM-DEMO:DEMODEMO))

**Comment by slyrus** [comment]
When I try to run that, I get:
```
Asynchronous REQUEST-ERROR in request 216 (last request was 261)  Code 214.0 [unknown]
   [Condition of type XLIB:REQUEST-ERROR]

Restarts:
 0: [CONTINUE] Ignore
 1: [RESTART-EVENT-LOOP] Restart CLIM's event loop.
 2: [ABORT] abort thread (#<THREAD "#<CLIM-CLX-FB::CLX-FB-PORT {100830A5A3}>'s event process." RUNNING {10085E3FC3}>)

Backtrace:
  0: (XLIB::READ-ERROR-INPUT #<XLIB:DISPLAY :0 (The X.Org Foundation R11804000)> 216 #S(XLIB::REPLY-BUFFER :SIZE 32 :IBUF8 #(0 1 216 0 28 0 ...) :NEXT #S(XLIB::REPLY-BUFFER :SIZE 32 :IBUF8 #(6 0 189 0 206 ..
  1: (XLIB::READ-INPUT #<XLIB:DISPLAY :0 (The X.Org Foundation R11804000)> 0 NIL #<FUNCTION (LAMBDA (XLIB:DISPLAY) :IN XLIB::WAIT-FOR-EVENT) {22F3020B}> #<XLIB:DISPLAY :0 (The X.Org Foundation R11804000)>)
  2: (XLIB::WAIT-FOR-EVENT #<XLIB:DISPLAY :0 (The X.Org Foundation R11804000)> 0 NIL)
  3: ((FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN XLIB:EVENT-LISTEN))
  4: ((FLET #:WITHOUT-INTERRUPTS-BODY-386 :IN SB-THREAD::CALL-WITH-RECURSIVE-LOCK))
  5: (SB-THREAD::CALL-WITH-RECURSIVE-LOCK #<CLOSURE (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN XLIB:EVENT-LISTEN) {817F24B}> #<SB-THREAD:MUTEX "CLX Event Lock" owner: #<SB-THREAD:THREAD "#<CLIM-CLX-FB:..
  6: ((FLET SB-IMPL::TIMEOUT-BODY :IN XLIB:EVENT-LISTEN))
  7: (XLIB:EVENT-LISTEN #<XLIB:DISPLAY :0 (The X.Org Foundation R11804000)> 0)
  8: ((:METHOD CLIM-BACKEND:GET-NEXT-EVENT (CLIM-CLX-FB::CLX-FB-PORT)) #<CLIM-CLX-FB::CLX-FB-PORT {100830A5A3}> :WAIT-FUNCTION #<unused argument> :TIMEOUT #<unused argument>) [fast-method]
  9: ((SB-PCL::EMF CLIM-BACKEND:GET-NEXT-EVENT) #<unused argument> #<unused argument> #<CLIM-CLX-FB::CLX-FB-PORT {100830A5A3}> :WAIT-FUNCTION NIL :TIMEOUT NIL)
 10: ((:METHOD CLIM:PROCESS-NEXT-EVENT (CLIM:BASIC-PORT)) #<CLIM-CLX-FB::CLX-FB-PORT {100830A5A3}> :WAIT-FUNCTION NIL :TIMEOUT NIL) [fast-method]
 11: ((LAMBDA NIL :IN CLIM-CLX::INITIALIZE-CLX))
 12: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
 13: ((FLET #:WITHOUT-INTERRUPTS-BODY-1159 :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
 14: ((FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
 15: ((FLET #:WITHOUT-INTERRUPTS-BODY-358 :IN SB-THREAD::CALL-WITH-MUTEX))
 16: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE) {817FCFB}> #<SB-THREAD:MUTEX "thread result lock" owner: #<SB-THREAD:THREAD "..
 17: (SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE #<SB-THREAD:THREAD "#<CLIM-CLX-FB::CLX-FB-PORT {100830A5A3}>'s event process." RUNNING {10085E3FC3}> NIL #<CLOSURE (LAMBDA NIL :IN BORDEAUX-THREADS::BIND..
 18: ("foreign function: call_into_lisp")
 19: ("foreign function: new_thread_trampoline")
```

**Comment by gas2serra** [comment]
Or "Drawing test" example (only render)

**Comment by gas2serra** [comment]
I'm testing it in another machine...

**Comment by gas2serra** [comment]
It works fine. 
May be you can try to clean your cache and to recompile all.
With the clx backend does it work?
Alessandro

**Comment by dkochmanski** [comment]
backend runs fine for me. I have dozen of these when running (nothing serious I think)
```
STYLE-WARNING:
   slot names with the same SYMBOL-NAME but different SYMBOL-PACKAGE (possible
   package problem) for class #<STANDARD-CLASS CLIM-CLX-FB::CLX-FB-PORT>:
       (CLIM-CLX::FONT-FAMILIES MCCLIM-RENDER::FONT-FAMILIES)
create class Clx-Fb-TOP-LEVEL-SHEET-PANE-DUMMY
STYLE-WARNING:
   slot names with the same SYMBOL-NAME but different SYMBOL-PACKAGE (possible
   package problem) for class #<STANDARD-CLASS CLIM-CLX-FB::CLX-FB-PORT>:
       (CLIM-CLX::FONT-FAMILIES MCCLIM-RENDER::FONT-FAMILIES)
STYLE-WARNING:
   slot names with the same SYMBOL-NAME but different SYMBOL-PACKAGE (possible
   package problem) for class #<STANDARD-CLASS CLIM-CLX-FB::CLX-FB-PORT>:
       (CLIM-CLX::FONT-FAMILIES MCCLIM-RENDER::FONT-FAMILIES)
STYLE-WARNING:
   slot names with the same SYMBOL-NAME but different SYMBOL-PACKAGE (possible
   package problem) for class #<STANDARD-CLASS CLIM-CLX-FB::CLX-FB-PORT>:
       (CLIM-CLX::FONT-FAMILIES MCCLIM-RENDER::FONT-FAMILIES)
STYLE-WARNING:
   slot names with the same SYMBOL-NAME but different SYMBOL-PACKAGE (possible
   package problem) for class #<STANDARD-CLASS CLIM-CLX-FB::CLX-FB-PORT>:
       (CLIM-CLX::FONT-FAMILIES MCCLIM-RENDER::FONT-FAMILIES)
create class Clx-Fb-MENU-UNMANAGED-TOP-LEVEL-SHEET-PANE-DUMMY
```

**Comment by gas2serra** [comment]
Style-warnings fixed.
@slyrus have you tried this patch with standard clx backend? It improves the performance of "Drawing test". I'm using sbcl 1.3.16 with the master branch of clx.

File: Extensions/render/fonts.lisp [hunk 1]
```diff
@@ -67,6 +85,29 @@
     (setf (svref cache hash) key-number
           (svref cache (+ 256 hash)) value)))
 
+;;;
+;;;
+;;;
+
+
```
Comment (dkochmanski) [inline Extensions/render/fonts.lisp]: do we need such big gap in the code?

File: Extensions/render/fonts.lisp [hunk 2]
```diff
@@ -56,7 +74,7 @@
 (declaim (inline gcache-get))
 
 (defun gcache-get (cache key-number)  
-  (declare (optimize (speed 3))
+  (declare ;;(optimize (speed 3))
            (type (simple-array t (512))))
```
Comment (dkochmanski) [inline Extensions/render/fonts.lisp]: why did you remove speed delcaration here? if it's bogus (i.e break things, either reduce it to lower level, or remove)

File: Extensions/render/fonts.lisp [hunk 3]
```diff
@@ -67,6 +85,29 @@
...
+			    (- dx) dy)
+			   (make-rectangle* 0 0 (* 1 width) (* 1 height))))
+    
+    ;;(save-image-to-file image "/tmp/a.png")
+    image))
+
+;;;
+;;;
+;;;
+
```
Comment (dkochmanski) [inline Extensions/render/fonts.lisp]: another big gap

File: Extensions/render/fonts.lisp [hunk 4]
```diff
@@ -76,7 +117,7 @@
   ;; -> (width ascent descent left right
   ;; font-ascent font-descent direction
   ;; first-not-done)  
-  (declare (optimize (speed 3))
+  (declare ;;(optimize (speed 3))
            (ignore translate))
```
Comment (dkochmanski) [inline Extensions/render/fonts.lisp]: see my previous comment wrt speed declaration

File: Extensions/render/fonts.lisp [hunk 5]
```diff
@@ -113,7 +154,7 @@
 
 (defmethod text-size ((medium render-medium-mixin) string
                       &key text-style (start 0) end)
-  (declare (optimize (speed 3)))
+  ;;(declare (optimize (speed 3)))
   (when (characterp string)
```
Comment (dkochmanski) [inline Extensions/render/fonts.lisp]: see my previous comment wrt speed declaration

File: Extensions/render/medium.lisp [hunk 6]
```diff
@@ -187,12 +232,32 @@
...
+						       
+						       (%medium-fill-image-mask
+							medium
+							opacity-image
+							min-x min-y
+							(- max-x min-x) (- max-y min-y)
+							(- (round x1)) (- (round y1))
+							)))))
+					       
+					       #+nil(%medium-fill-paths medium paths t transformation)
```
Comment (dkochmanski) [inline Extensions/render/medium.lisp]: please use `#+(or)`

File: Extensions/render/medium.lisp [hunk 7]
```diff
@@ -187,12 +232,32 @@
...
 		    (:bottom (+ y (- baseline text-height))))))
-	(multiple-value-bind (x y)
-	    (transform-position (sheet-native-transformation
-				   (medium-sheet medium))
-				x y)
-	  (let ((paths (string-primitive-paths x y string xfont size)))
-	    (%medium-fill-paths medium paths nil)))))))
+	(let ((paths (string-primitive-paths x y string xfont size
+					     (lambda (paths opacity-image dx dy transformation)
+					       (let ((msheet (sheet-mirrored-ancestor (medium-sheet medium))))
```
Comment (dkochmanski) [inline Extensions/render/medium.lisp]: for such big function I think that it would be good to factor it into toplevel helper function

File: Extensions/render/mirror.lisp [hunk 8]
```diff
@@ -52,21 +50,93 @@
...
   (with-slots (dirty-region) mirror
     (if dirty-region
 	(setf dirty-region (region-union dirty-region region))
 	(setf dirty-region region))))
 
-;;; utility function
+;;;
+;;;
+;;;
+
```
Comment (dkochmanski) [inline Extensions/render/mirror.lisp]: unjustified empty space

File: Extensions/render/mirror.lisp [hunk 9]
```diff
@@ -52,21 +50,93 @@
...
+    (with-slots (image-lock) mirror
+      (climi::with-lock-held (image-lock)
+	(call-next-method)))))
+
+(defmethod %draw-image ((mirror image-mirror-mixin)
+			image x y width height
+			dst-dx dst-dy
+			clip-region)
+  (let ((region #+nil(rgb-image-copy
+		 image
```
Comment (dkochmanski) [inline Extensions/render/mirror.lisp]: please use `#+(or)` or remove this. Also indent the block, it's not obviuos at a glympse, where rgb-image-copy block ends.

File: Extensions/render/mirror.lisp [hunk 10]
```diff
@@ -52,21 +50,93 @@
...
+		(copy-image
+		 (image-mirror-image mirror)
+		 image
+		 :x x :y y :width width :height height
+		 ;;:x-dst x-dest :y-dst y-dest
+		 :src-dx dst-dx
+		 :src-dy dst-dy
+		 ;;:clip-region clip-region
+		 )
+		))
```
Comment (dkochmanski) [inline Extensions/render/mirror.lisp]: are these commented arguments valid, or just leftovers? also please close blocks in the same line, like `)))`

File: Extensions/render/opticl-image.lisp [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+;;;
+
+;;;
+;;; macro to build drawing function
+;;;
+
+
+
+
+
```
Comment (dkochmanski) [inline Extensions/render/opticl-image.lisp]: unjustified empty space. If you want to spearate this parts more visibly, you may add page break, or a comment box

File: Extensions/render/rgba-image.lisp [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+;;;
+
+;;;
+;;; macro to build drawing function
+;;;
+
+
+
+
+
```
Comment (dkochmanski) [inline Extensions/render/rgba-image.lisp]: huge empty space

**Comment by gas2serra** [comment]
Hi Daniel, 
ok, I'll improve the style.

**Comment by dkochmanski** [comment]
thanks

**Comment by dkochmanski** [comment]
Looks good, thanks! Merged.
